arch/riscv: Enable NMI and regular interrupts/traps when the hardware includes SMRNMI extension

AFOliveira opened the pull request:
Add RISC-V SMRNMI Extension Support to Zephyr

This PR adds support for the RISC-V SMRNMI (Resumable Non-Maskable Interrupt) extension.

Changes:
- Add CONFIG_RISCV_ISA_EXT_SMRNMI Kconfig option
- Enable NMIE bit in mnstatus CSR during boot when configured

The NMIE (NMI Enable) bit controls whether the CPU receives non-maskable interrupts. It defaults to 0 after reset, meaning NMIs are generated by hardware but never delivered to the CPU. This causes critical events like watchdog timeouts, ECC errors, and bus faults to be silently ignored.

Setting NMIE=1 during boot ensures the CPU can receive and handle these critical interrupts. This fixes both production issues where hardware events were missed and twister test failures on SMRNMI platforms.

Without this change, SMRNMI hardware cannot deliver any NMIs despite generating them, leaving systems vulnerable to undetected hardware failures.

npitre reviewed (commented):
Please define `mnstatus` in include/zephyr/arch/riscv/csr.h.

AFOliveira commented:
@npitre Sure, thanks for the help!

AFOliveira commented:
@npitre refactored as requested, Added the rest of the extensions CSRs for completeness.

AFOliveira committed:
arch/riscv: Enable NMI delivery for SMRNMI hardware

Add option to enable NMI delivery on boot for SMRNMI hardware.

Changes:
- Add CONFIG_RISCV_SMRNMI_ENABLE_NMI_DELIVERY Kconfig option
- Define SMRNMI CSRs in arch/riscv/include/csr.h
- Set NMIE bit during boot to enable NMI delivery

SMRNMI hardware generates but doesn't deliver NMIs when NMIE=0 (default).
This causes twister test failures and prevents handling of critical
hardware events like watchdog NMIs and ECC errors.

Setting NMIE=1 enables NMI delivery, but note that this implementation
only sets the enable bit - it does not provide full SMRNMI support
(no mnret instruction handling, no RNMI handlers). Users must implement
proper RNMI handlers in SoC-specific code to avoid undefined behavior.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

fkokosinski commented on arch/riscv/Kconfig.isa:
```diff
@@ -99,6 +99,17 @@ config RISCV_ISA_EXT_C
 	  which reduces static and dynamic code size by adding short 16-bit
 	  instruction encodings for common operations.
 
+config RISCV_ISA_EXT_SMRNMI
```
This extension should probably select `RISCV_ISA_EXT_ZICSR`.

AFOliveira replied:
Indeed, it does, thanks for the catch!

fkokosinski reviewed (commented):
Hey, thanks for the PR. I've got two additional questions:
1. Does the binutils package that is provided by Zephyr SDK come in a version where the smrnmi extension is already present?
2. Shouldn't we modify the cmake script in `cmake/compiler/gcc/target_riscv.cmake` so that smrnmi is provided to GCC by the march argument?

fkokosinski commented:
@AFOliveira -

Thanks for answering my questions. I guess my concerns boil down to a single question: how would a Zephyr user/developer interpret the `CONFIG_RISCV_ISA_EXT_SMRNMI` option?

The Kconfig prompt, symbol name and help string are worded in such a way  the user would expect the extension to be fully supported, including the `mnret` instruction, meaning - the user is being misled here.

My suggestion would be to rename this option so that it reflects what it actually does: it sets the `MNSTATUS.MNIE` field to 1.

AFOliveira commented:
> The Kconfig prompt, symbol name and help string are worded in such a way the user would expect the extension to be fully supported, including the `mnret` instruction, meaning - the user is being misled here.
> 
> My suggestion would be to rename this option so that it reflects what it actually does: it sets the `MNSTATUS.MNIE` field to 1

@fkokosinski  I did just what you proposed and changed the Kconfig option name. I plan to then introduce SoC specific code to handle the RNMIs, but at this point the spec is too wide for me to do what I intended in first place - set a handler address.

ruuddw commented:
Note that the main reason for this simple, minimal PR is this (from the RV extension spec):

> mnstatus also holds the NMIE bit. When NMIE=1, nonmaskable interrupts are enabled. **When NMIE=0, all interrupts are disabled**.

Goal of this PR is not to provide RNMI support, but to still have regular interrupts/traps when the hardware includes this feature.
